Dockerfile

```diff
@@ -163,47 +164,67 @@ RUN mkdir -p ${GOPATH}/src/github.com/chrusty/protoc-gen-jsonschema && \
 
 
 FROM rust:${RUST_VERSION}-alpine as rust_builder
+ARG TARGETARCH
 RUN apk add --no-cache curl
-RUN rustup target add x86_64-unknown-linux-musl
+RUN case ${TARGETARCH} in \
+         "amd64")  RUST_TARGET=x86_64-unknown-linux-musl  ;; \
+         "arm64")  RUST_TARGET=aarch64-unknown-linux-musl ;; \
+         *)        echo "ERROR: Machine arch ${TARGETARCH} not supported." ;; \
+    esac && \
+    rustup target add $RUST_TARGET
 
 ARG RUST_PROTOBUF_VERSION
-RUN mkdir -p /rust-protobuf && \
+RUN case ${TARGETARCH} in \
+         "amd64")  RUST_TARGET=x86_64-unknown-linux-musl  ;; \
+         "arm64")  RUST_TARGET=aarch64-unknown-linux-musl ;; \
+    esac && \
+    mkdir -p /rust-protobuf && \
     curl -sSL https://api.github.com/repos/stepancheg/rust-protobuf/tarball/v${RUST_PROTOBUF_VERSION} | tar xz --strip 1 -C /rust-protobuf && \
-    cd /rust-protobuf/protobuf-codegen && cargo build --target=x86_64-unknown-linux-musl --release && \
-    install -Ds /rust-protobuf/target/x86_64-unknown-linux-musl/release/protoc-gen-rust /out/usr/bin/protoc-gen-rust
+    cd /rust-protobuf/protobuf-codegen && cargo build --target=$RUST_TARGET --release && \
+    install -Ds /rust-protobuf/target/$RUST_TARGET/release/protoc-gen-rust /out/usr/bin/protoc-gen-rust
 
 ARG GRPC_RUST_VERSION
-RUN mkdir -p /grpc-rust && curl -sSL https://api.github.com/repos/stepancheg/grpc-rust/tarball/v${GRPC_RUST_VERSION} | tar xz --strip 1 -C /grpc-rust && \
-    cd /grpc-rust/grpc-compiler && cargo build --target=x86_64-unknown-linux-musl --release && \
-    install -Ds /grpc-rust/target/x86_64-unknown-linux-musl/release/protoc-gen-rust-grpc /out/usr/bin/protoc-gen-rust-grpc
+RUN case ${TARGETARCH} in \
+         "amd64")  RUST_TARGET=x86_64-unknown-linux-musl  ;; \
```

Comment:
nit; I wish there was a way to only do this matching once

Reply:
Because each docker layer is created using a fresh shell environment, I believe the only way to achieve this is to write the previous result of the case statement to a file and then load it from the file when needed in later commands - it's not possible to set ARG or ENV values as the result of a command.